core/src/ldml/ldml_processor.cpp

```diff
@@ -168,21 +222,83 @@ ldml_processor::process_event(
 
     switch (vk) {
     case KM_KBP_VKEY_BKSP:
-      state->context().pop_back();
-      state->actions().push_backspace(KM_KBP_BT_UNKNOWN); // Assuming we don't know the character
+      {
+        KMX_DWORD last_char = 0UL; // Assuming we don't know the character
+        // attempt to get the last char
+        auto end = state->context().rbegin();
+        if(end != state->context().rend()) {
+          if((*end).type == KM_KBP_CT_CHAR) {
+            last_char = (*end).character;
+          }
+        }
+        if (last_char == 0UL) {
+          state->actions().push_backspace(KM_KBP_BT_UNKNOWN);
+        } else {
+          state->actions().push_backspace(KM_KBP_BT_CHAR, last_char);
+        }
+        state->context().pop_back();
+      }
       break;
     default:
+      // from kmx_processor.cpp
+      // Construct a context buffer from the items up until the last KM_KBP_CT_MARKER marker
+      ldml::string_list ctxt;
+      auto cp = state->context();
+      // We're only interested in as much of the context as is a KM_KBP_CT_CHAR.
+      // This will stop at the KM_KBP_CT_MARKER type.
+      uint8_t last_type = KM_KBP_BT_UNKNOWN;
+      for (auto c = cp.rbegin(); c != cp.rend(); c++) {
+        last_type = c->type;
+        if (last_type != KM_KBP_BT_CHAR) {
+          break;
+        }
+        km::kbp::kmx::char16_single buf;
+        const int len = km::kbp::kmx::Utf32CharToUtf16(c->character, buf);
+        const std::u16string str(buf.ch, len);
+        ctxt.push_front(str); // prepend to string
+      }
+      if (last_type != KM_KBP_BT_MARKER) {
+        // There was no beginning-of-translation marker.
+        //Add one.
+        state->context().push_marker(0x0);
```

Comment:
I'd have expected we'd push `KM_KBP_BT_MARKER` or something like that.

Reply:
This is its own marker…